Source/NSOrderedSet.m

```diff
@@ -0,0 +1,1814 @@
+{
+  return [[self objectsAtIndexes: indexSet]
+	   indexesOfObjectsWithOptions: opts
+			   passingTest: predicate];
+}
+
+- (NSIndexSet *) indexesOfObjectsPassingTest:(GSPredicateBlock)predicate
+{
+  return [self indexesOfObjectsWithOptions: 0 passingTest: predicate];
+}
+
+- (NSIndexSet *) indexesOfObjectsWithOptions:(NSEnumerationOptions)opts
+                            passingTest:(GSPredicateBlock)predicate
+{
+  /* TODO: Concurrency. */
+  NSMutableIndexSet     *set = [NSMutableIndexSet indexSet];
+  BLOCK_SCOPE BOOL      shouldStop = NO;
+  id<NSFastEnumeration> enumerator = self;
+  NSUInteger            count = 0;
+  BLOCK_SCOPE NSLock    *setLock = nil;
+
+  /* If we are enumerating in reverse, use the reverse enumerator for fast
+   * enumeration. */
+  if (opts & NSEnumerationReverse)
+    {
+      enumerator = [self reverseObjectEnumerator];
+    }
+  if (opts & NSEnumerationConcurrent)
+    {
+      setLock = [NSLock new];
+    }
+  {
+    GS_DISPATCH_CREATE_QUEUE_AND_GROUP_FOR_ENUMERATION(enumQueue, opts)
+    FOR_IN (id, obj, enumerator)
+#     if __has_feature(blocks) && (GS_USE_LIBDISPATCH == 1)
+
+      dispatch_group_async(enumQueueGroup, enumQueue, ^(void){
+        if (shouldStop)
+        {
+	  return;
+        }
+        if (predicate(obj, count, &shouldStop))
+        {
+	  [setLock lock];
+	  [set addIndex: count];
+	  [setLock unlock];
+        }
+      });
+#     else
+      if (CALL_BLOCK(predicate, obj, count, &shouldStop))
+        {
+	  /* TODO: It would be more efficient to collect an NSRange and only
+	   * pass it to the index set when CALL_BLOCK returned NO. */
+	  [set addIndex: count];
+        }
+#     endif
+      if (shouldStop)
+        {
+	  break;
+        }
+      count++;
+    END_FOR_IN(enumerator)
+    GS_DISPATCH_TEARDOWN_QUEUE_AND_GROUP_FOR_ENUMERATION(enumQueue, opts);
+  }
+  RELEASE(setLock);
+  return set;
+}
+
+- (NSEnumerator *) objectEnumerator
+{
+  [self subclassResponsibility: _cmd];
+  return nil;
+}
+
+- (NSEnumerator *) reverseObjectEnumerator
+{
+    [self subclassResponsibility: _cmd];
+    return nil;
+}
+
+- (NSOrderedSet *)reversedOrderedSet
+{
+  NSEnumerator *e = [self reverseObjectEnumerator];
+  NSMutableArray *a = [NSMutableArray arrayWithCapacity: [self count]];
+  id o = nil;
+
+  // Build the reverse array...
+  while ((o = [e nextObject]) != nil)
+    {
+      [a addObject: o];
+    }
+
+  // Create and return reverse ordered set...
+  return [NSOrderedSet orderedSetWithArray: a];
+}
+
+- (void) getObjects: (__unsafe_unretained id[])aBuffer range: (NSRange)aRange
+{
+  NSUInteger i, j = 0, c = [self count], e = aRange.location + aRange.length;
+  IMP	get = [self methodForSelector: oaiSel];
+
+  GS_RANGE_CHECK(aRange, c);
+
+  for (i = aRange.location; i < e; i++)
+    aBuffer[j++] = (*get)(self, oaiSel, i);
+}
+ 
+// Key value coding support
+- (void) setValue: (id)value forKey: (NSString*)key
+{
+  NSUInteger    i;
+  NSUInteger	count = [self count];
```

Comment:
Why didn't you use an enumerator as in the method below? I always prefer to express the same idea in a similar way in different places.

Reply:
Done.